File: ios/Classes/FsManager.swift
import Foundation
import iOSMcuManagerLibrary

class FsManager {
    let transport: McuMgrBleTransport
    let fs: FileSystemManager
    let peripheral: CBPeripheral

    init(peripheral: CBPeripheral) {
        self.peripheral = peripheral
        self.transport = McuMgrBleTransport(peripheral)
        self.fs = FileSystemManager(transporter: self.transport)
    }

    func list(path: String, completion: @escaping (Data?, Error?) -> Void) {
        fs.list(path: path) { response, error in
            var proto = ProtoFsListResponse()
            if let err = error {
                proto.success = false
                proto.error = ProtoError(localizedDescription: err.localizedDescription)
            } else if let resp = response {
                proto.success = true
                proto.files = resp.map { entry in
                    var e = ProtoFsEntry()
                    e.name = entry.name
                    e.size = UInt64(entry.size)
                    e.dir = entry.isDir
                    e.mode = UInt32(entry.mode)
                    e.mtime = UInt64(entry.mtime)
                    return e
                }
            } else {
                proto.success = false
            }

            do {
                completion(try proto.serializedData(), nil)
            } catch {
                completion(nil, error)
            }
        }
    }

    func stat(path: String, completion: @escaping (Data?, Error?) -> Void) {
        fs.stat(path: path) { response, error in
            var proto = ProtoFsStatResponse()
            if let err = error {
                proto.success = false
                proto.error = ProtoError(localizedDescription: err.localizedDescription)
            } else if let resp = response {
                proto.success = true
                var e = ProtoFsEntry()
                e.name = resp.name
                e.size = UInt64(resp.size)
                e.dir = resp.isDir
                e.mode = UInt32(resp.mode)
                e.mtime = UInt64(resp.mtime)
                proto.stats = e
            } else {
                proto.success = false
            }

            do {
                completion(try proto.serializedData(), nil)
            } catch {
                completion(nil, error)
            }
        }
    }

    func remove(path: String, completion: @escaping (Data?, Error?) -> Void) {
        fs.remove(path: path) { response, error in
            var proto = ProtoFsRemoveResponse()
            if let err = error {
                proto.success = false
                proto.error = ProtoError(localizedDescription: err.localizedDescription)
            } else if let resp = response {
                proto.success = resp
            } else {
                proto.success = false
            }

            do {
                completion(try proto.serializedData(), nil)
            } catch {
                completion(nil, error)
            }
        }
    }

    func open(path: String, flags: Int32, completion: @escaping (Data?, Error?) -> Void) {
        fs.open(path: path, flags: flags) { response, error in
            var proto = ProtoFsOpenResponse()
            if let err = error {
                proto.success = false
                proto.error = ProtoError(localizedDescription: err.localizedDescription)
            } else if let resp = response {
                proto.success = true
                proto.fd = Int32(resp)
            } else {
                proto.success = false
            }

            do {
                completion(try proto.serializedData(), nil)
            } catch {
                completion(nil, error)
            }
        }
    }

    func read(fd: Int32, offset: Int32, length: Int32, completion: @escaping (Data?, Error?) -> Void) {
        fs.read(fd: fd, offset: offset, length: length) { response, error in
            var proto = ProtoFsReadResponse()
            if let err = error {
                proto.success = false
                proto.error = ProtoError(localizedDescription: err.localizedDescription)
            } else if let resp = response {
                proto.success = true
                proto.data = resp
            } else {
                proto.success = false
            }

            do {
                completion(try proto.serializedData(), nil)
            } catch {
                completion(nil, error)
            }
        }
    }

    func write(fd: Int32, data: Data, offset: Int32, completion: @escaping (Data?, Error?) -> Void) {
        fs.write(fd: fd, data: data, offset: offset) { response, error in
            var proto = ProtoFsWriteResponse()
            if let err = error {
                proto.success = false
                proto.error = ProtoError(localizedDescription: err.localizedDescription)
            } else if let resp = response {
                proto.success = true
                proto.written = UInt32(resp)
            } else {
                proto.success = false
            }

            do {
                completion(try proto.serializedData(), nil)
            } catch {
                completion(nil, error)
            }
        }
    }

    func close(fd: Int32, completion: @escaping (Data?, Error?) -> Void) {
        fs.close(fd: fd) { response, error in
            var proto = ProtoFsCloseResponse()
            if let err = error {
                proto.success = false
                proto.error = ProtoError(localizedDescription: err.localizedDescription)
            } else if let resp = response {
                proto.success = resp
            } else {
                proto.success = false
            }

            do {
                completion(try proto.serializedData(), nil)
            } catch {
                completion(nil, error)
            }
        }
    }
}

File: android/src/main/kotlin/no/nordicsemi/android/mcumgr_flutter/FsManager.kt
package no.nordicsemi.android.mcumgr_flutter

import android.util.Log
import com.google.protobuf.kotlin.toByteString
import io.runtime.mcumgr.McuMgrCallback
import io.runtime.mcumgr.exception.McuMgrException
import io.runtime.mcumgr.managers.FileSystemManager
import io.runtime.mcumgr.response.fs.McuMgrFsListResponse
import io.runtime.mcumgr.response.fs.McuMgrFsStatResponse
import no.nordicsemi.android.mcumgr_flutter.gen.*
import no.nordicsemi.android.mcumgr_flutter.utils.StreamHandler

class FsManager(
    private val transport: LoggableMcuMgrBleTransport,
    private val readStreamHandler: StreamHandler,
    private val logStreamHandler: StreamHandler
) {
    private val fsManager: FileSystemManager = FileSystemManager(transport)
    private val address: String = transport.bluetoothDevice.address

    fun list(path: String, callbackResult: (ByteArray) -> Unit) {
        val cb = object : McuMgrCallback<McuMgrFsListResponse> {
            override fun onResponse(response: McuMgrFsListResponse) {
                val entries = response.files?.map { file ->
                    ProtoFsEntry(
                        name = file.name,
                        size = file.size,
                        dir = file.type == io.runtime.mcumgr.fs.McuMgrFsEntry.Type.DIRECTORY,
                        mode = file.mode.toInt(),
                        mtime = file.mtime
                    )
                } ?: listOf()

                val res = ProtoFsListResponse(
                    uuid = address,
                    files = entries,
                    success = true
                )
                callbackResult(res.encode())
            }

            override fun onError(exception: McuMgrException) {
                val res = ProtoFsListResponse(
                    uuid = address,
                    success = false,
                    error = ProtoError(localizedDescription = exception.message ?: "Unknown error")
                )
                callbackResult(res.encode())
            }
        }
        fsManager.list(path, cb)
    }

    fun stat(path: String, callbackResult: (ByteArray) -> Unit) {
        val cb = object : McuMgrCallback<McuMgrFsStatResponse> {
            override fun onResponse(response: McuMgrFsStatResponse) {
                val entry = response.entry?.let { e ->
                    ProtoFsEntry(
                        name = e.name,
                        size = e.size,
                        dir = e.type == io.runtime.mcumgr.fs.McuMgrFsEntry.Type.DIRECTORY,
                        mode = e.mode.toInt(),
                        mtime = e.mtime
                    )
                }
                val res = ProtoFsStatResponse(
                    uuid = address,
                    stats = entry,
                    success = true
                )
                callbackResult(res.encode())
            }

            override fun onError(exception: McuMgrException) {
                val res = ProtoFsStatResponse(
                    uuid = address,
                    success = false,
                    error = ProtoError(localizedDescription = exception.message ?: "Unknown error")
                )
                callbackResult(res.encode())
            }
        }

        fsManager.stat(path, cb)
    }

    fun remove(path: String, callbackResult: (ByteArray) -> Unit) {
        val cb = object : McuMgrCallback<Void> {
            override fun onResponse(response: Void?) {
                val res = ProtoFsRemoveResponse(
                    uuid = address,
                    success = true
                )
                callbackResult(res.encode())
            }

            override fun onError(exception: McuMgrException) {
                val res = ProtoFsRemoveResponse(
                    uuid = address,
                    success = false,
                    error = ProtoError(localizedDescription = exception.message ?: "Unknown error")
                )
                callbackResult(res.encode())
            }
        }
        fsManager.remove(path, cb)
    }

    fun open(path: String, flags: Int, callbackResult: (ByteArray) -> Unit) {
        val cb = object : McuMgrCallback<Int> {
            override fun onResponse(response: Int) {
                val res = ProtoFsOpenResponse(
                    uuid = address,
                    fd = response,
                    success = true
                )
                callbackResult(res.encode())
            }

            override fun onError(exception: McuMgrException) {
                val res = ProtoFsOpenResponse(
                    uuid = address,
                    fd = -1,
                    success = false,
                    error = ProtoError(localizedDescription = exception.message ?: "Unknown error")
                )
                callbackResult(res.encode())
            }
        }
        fsManager.open(path, flags, cb)
    }

    fun read(fd: Int, offset: Long, length: Long, callbackResult: (ByteArray) -> Unit) {
        val cb = object : McuMgrCallback<io.runtime.mcumgr.response.fs.McuMgrFsReadResponse> {
            override fun onResponse(response: io.runtime.mcumgr.response.fs.McuMgrFsReadResponse) {
                val data = response.data
                val res = ProtoFsReadResponse(
                    uuid = address,
                    data = data.toByteArray(),
                    done = response.eof,
                    success = true
                )
                callbackResult(res.encode())
            }

            override fun onError(exception: McuMgrException) {
                val res = ProtoFsReadResponse(
                    uuid = address,
                    data = ByteArray(0),
                    done = true,
                    success = false,
                    error = ProtoError(localizedDescription = exception.message ?: "Unknown error")
                )
                callbackResult(res.encode())
            }
        }
        fsManager.read(fd, offset, length, cb)
    }

    fun write(fd: Int, data: ByteArray, offset: Long, callbackResult: (ByteArray) -> Unit) {
        val cb = object : McuMgrCallback<Long> {
            override fun onResponse(response: Long) {
                val res = ProtoFsWriteResponse(
                    uuid = address,
                    written = response,
                    success = true
                )
                callbackResult(res.encode())
            }

            override fun onError(exception: McuMgrException) {
                val res = ProtoFsWriteResponse(
                    uuid = address,
                    written = 0,
                    success = false,
                    error = ProtoError(localizedDescription = exception.message ?: "Unknown error")
                )
                callbackResult(res.encode())
            }
        }
        fsManager.write(fd, data, offset, cb)
    }

    fun close(fd: Int, callbackResult: (ByteArray) -> Unit) {
        val cb = object : McuMgrCallback<Void> {
            override fun onResponse(response: Void?) {
                val res = ProtoFsCloseResponse(
                    uuid = address,
                    success = true
                )
                callbackResult(res.encode())
            }

            override fun onError(exception: McuMgrException) {
                val res = ProtoFsCloseResponse(
                    uuid = address,
                    success = false,
                    error = ProtoError(localizedDescription = exception.message ?: "Unknown error")
                )
                callbackResult(res.encode())
            }
        }
        fsManager.close(fd, cb)
    }
}
